router/h2/src/main/scala/io/buoyant/router/h2/StreamStatsFilter.scala

```diff
@@ -1,83 +1,130 @@
 package io.buoyant.router.h2
 
 import com.twitter.finagle.{Status => _, _}
-import com.twitter.finagle.buoyant.h2.{Request, Response, Stream}
+import com.twitter.finagle.buoyant.h2.{Frame, H2ReqRep, H2ResponseClassifier, Request, Response, Stream, param => h2param}
+import com.twitter.finagle.buoyant.h2.StreamProxy._
+import com.twitter.finagle.service.ResponseClass.Successful
 import com.twitter.finagle.stats.StatsReceiver
 import com.twitter.util._
 
 object StreamStatsFilter {
   val role = Stack.Role("StreamStatsFilter")
   val module: Stackable[ServiceFactory[Request, Response]] =
-    new Stack.Module1[param.Stats, ServiceFactory[Request, Response]] {
+    new Stack.Module2[param.Stats, h2param.H2ResponseClassifier, ServiceFactory[Request, Response]] {
       override def role = StreamStatsFilter.role
       override def description = "Record stats on h2 streams"
-      override def make(statsP: param.Stats, next: ServiceFactory[Request, Response]) = {
+      override def make(
+        statsP: param.Stats,
+        classifierP: h2param.H2ResponseClassifier,
+        next: ServiceFactory[Request, Response]
+      ) = {
         val param.Stats(stats) = statsP
-        new StreamStatsFilter(stats).andThen(next)
+        val h2param.H2ResponseClassifier(classifier) = classifierP
+        new StreamStatsFilter(stats, classifier).andThen(next)
       }
     }
 }
 
-class StreamStatsFilter(statsReceiver: StatsReceiver)
+class StreamStatsFilter(statsReceiver: StatsReceiver, classifier: H2ResponseClassifier)
   extends SimpleFilter[Request, Response] {
 
-  class StreamStats(
-    stats: StatsReceiver,
-    durationName: Option[String] = None,
-    successName: Option[String] = None
-  ) {
+  class StreamStats(protected val stats: StatsReceiver, durationName: Option[String] = None) {
 
     private[this] val durationMs =
       stats.stat(s"${durationName.getOrElse("stream_duration")}_ms")
-    private[this] val successes =
-      stats.counter(s"${successName.getOrElse("stream_")}successes")
-    private[this] val failures =
-      stats.counter(s"${successName.getOrElse("stream_")}failures")
+    private[this] val successes = stats.counter("stream_successes")
+    private[this] val failures = stats.counter("stream_failures")
 
-    @inline def apply(startT: Stopwatch.Elapsed)(result: Try[_]): Unit = {
+    @inline def onEnd(startT: Stopwatch.Elapsed)(result: Try[_]): Unit = {
       durationMs.add(startT().inMillis)
       result match {
         case Return(_) => successes.incr()
         case Throw(_) => failures.incr()
       }
     }
+
+    @inline def apply(startT: Stopwatch.Elapsed)(result: Try[_]): Unit = onEnd(startT)(result)
```

Comment:
Oh, I think I see, you override apply in the subclass.

I would keep this all in apply and then in the subclass you can override it and call out to super.apply.

OR you can embrace "composition over inheritance" and have FrameStats contain a StreamStats instead of extending it.